Fix part of #13956: Using Python3 urlrequest

AmanJolly16 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13956 
2. This PR does the following: Changes all the occurrences of python_utils.url_request to its Python3 equivalent.
## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct


https://user-images.githubusercontent.com/61148168/144871222-652c02a1-e4df-4154-807f-f8852a2e2596.mov



<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

AmanJolly16 committed:
Updated the divide and url_request methods

AmanJolly16 committed:
Updated the divide and url_request methods

AmanJolly16 committed:
Refactor Python2 to Python3

AmanJolly16 committed:
Refactor Python2 to Python3

AmanJolly16 committed:
Refactor Python2 to Python3

AmanJolly16 committed:
Refactor Python2 to Python3

DubeySandeep commented:
@AmanJolly16 Thanks for the PR, we are expecting to remove the functions defined in python_utils instead of updating it! You can check the following PRs for reference: #13882, #13955

Also, make sure to follow [OppiaBot's comment](https://github.com/oppia/oppia/pull/14303#issuecomment-976489136)!

srijanreddy98 committed:
Merge branch 'develop' into utils-ref

AmanJolly16 committed:
Refactor

AmanJolly16 committed:
Refactor

AmanJolly16 committed:
Refactor

vojtechjelinek commented:
Hey @AmanJolly16 do you need help with anything?

AmanJolly16 committed:
Refactor

AmanJolly16 committed:
Refactor

AmanJolly16 committed:
Merge branch 'utils-ref' of https://github.com/AmanJolly16/oppia into utils-ref

AmanJolly16 committed:
Merge branch 'develop' into utils-ref

AmanJolly16 commented:
Hey @vojtechjelinek, sorry for the late reply. I replaced all the occurrences of python_utils.url_request. Is this fine now?
Also, I have another function "divide" assigned to me, I am not really able to find a suitable equivalent function for it in Python3. Can you help with that? Also, I am not really sure of how to (and what to) put in the "Proof that changes are correct" section.

AmanJolly16 committed:
Refactor

AmanJolly16 committed:
Refactor

AmanJolly16 committed:
Refactor

vojtechjelinek commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -126,7 +126,7 @@ def send_email_to_recipients(
         # https://docs.python.org/3.7/library/urllib.request.html#urllib-examples
         encoded_url = urllib.parse.urlencode(data).encode('ascii')
-        req = python_utils.url_request(server, encoded_url, header)
+        req = urlrequest.Request(server, encoded_url, header)
```
```suggestion
        req = urllib.request.Request(server, encoded_url, header)
```

vojtechjelinek commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -20,9 +20,9 @@
 
 import base64
 import urllib
+import urllib.request as urlrequest
```
Not needed (see the change below).

vojtechjelinek commented on scripts/check_if_pr_is_low_risk.py:
```diff
@@ -138,7 +138,7 @@ def lookup_pr(owner, repo, pull_number):
         200.
     """
-    request = python_utils.url_request(
+    request = urlrequest.Request(
```
```suggestion
    request = urllib.request.Request(
```

vojtechjelinek commented on scripts/check_if_pr_is_low_risk.py:
```diff
@@ -25,8 +25,8 @@
 import json
 import re
 import sys
+import urllib.request as urlrequest
```
Not needed.

vojtechjelinek commented on scripts/install_third_party.py:
```diff
@@ -142,7 +142,7 @@ def download_and_unzip_files(
 
             # Some downloads (like jqueryui-themes) may require a user-agent.
-            req = python_utils.url_request(source_url, None, {})
+            req = urlrequest.Request(source_url, None, {})
```
```suggestion
            req = urllib.request.Request(source_url, None, {})
```

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

Regarding divide, I recommend reading through the code and its documentation and trying to understand how should it work.

Regarding proof of changes, please start the server locally and record it, so that we can see that it can properly start.

AmanJolly16 committed:
Refactor

AmanJolly16 commented:
@vojtechjelinek after making the changes that you asked me to, during lint checks I am getting a "Please use python_utils.url_request()" error in the files in which you suggested changes. What do I do about this? Also, there is a test in oppia/core/platform/email/mailgun_email_services_test.py in which the test is trying to access the python_utils.py file and 
specifically the url_request function in it (more specifically line 69 is trying to access).

vojtechjelinek commented:
> @vojtechjelinek after making the changes that you asked me to, during lint checks I am getting a "Please use python_utils.url_request()" error in the files in which you suggested changes. What do I do about this? Also, there is a test in oppia/core/platform/email/mailgun_email_services_test.py in which the test is trying to access the python_utils.py file and specifically the url_request function in it (more specifically line 69 is trying to access).

To fix the lint checks please remove these lines https://github.com/oppia/oppia/blob/develop/scripts/linters/general_purpose_linter.py#L212-L217

In _mailgun_email_services_test.py_ you will need to swap the `Response` directly instead of swapping the removed function. The swap should be smth like this:
```python
swap_request_context = self.swap(urllib.request, 'Request', lambda *args: args)
```

Also, please DO NOT resolve comments from reviewers yourselves but keep it to the reviewer to decide whether the comment was resolved.

AmanJolly16 committed:
Lint and mailgun updated

AmanJolly16 commented:
@vojtechjelinek I am sorry I wasn't aware of that. I will make sure not to do it again. I have made the changes you asked me to, please have a look

vojtechjelinek commented on core/platform/email/mailgun_email_services_test.py:
```diff
@@ -17,9 +17,9 @@
 """Tests for the Mailgun API wrapper."""
 
 from __future__ import annotations
+import urllib
```
```suggestion

import urllib
```

vojtechjelinek reviewed (changes requested):
Thanks! Left one comment.

vojtechjelinek commented:
@AmanJolly16 Please add the proof of changes.

AmanJolly16 committed:
Update core/platform/email/mailgun_email_services_test.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

AmanJolly16 commented:
@vojtechjelinek Updated! Please have a look. Also, a lint test case was failing because we changed the url_request lint check. I have removed that test case. Is that fine?

AmanJolly16 committed:
Removed test case of url_request lint

AmanJolly16 committed:
Merge branch 'utils-ref' of https://github.com/AmanJolly16/oppia into utils-ref

vojtechjelinek reviewed (approved):
LGTM!
